crypto/bp/bp_g2.c

```diff
@@ -0,0 +1,1212 @@
+        FP2_zero(r->Z);
+        r->Z_is_one = 0;
+        return 1;
+    }
+
+    if (ctx == NULL && (ctx = new_ctx = BN_CTX_new()) == NULL)
+        return 0;
+
+    BN_CTX_start(ctx);
+    if ((t0 = FP2_new()) == NULL || (t1 = FP2_new()) == NULL
+        || (t2 = FP2_new()) == NULL || (t3 = FP2_new()) == NULL)
+        goto err;
+
+    /*
+     * Note that in this function we must not read components of 'a' once we
+     * have written the corresponding components of 'r'. ('r' might the same
+     * as 'a'.)
+     */
+
+    /*
+     * t1
+     */
+    if (!FP2_sqr(group, t0, a->X, ctx))
+        goto err;
+    if (!FP2_dbl(group, t1, t0))
+        goto err;
+    if (!FP2_add(group, t1, t1, t0))
+        goto err;
+    /*
+     * t1 = 3 * X_a^2 + a_curve * Z_a^4
+     */
+
+    /*
+     * Z_r
+     */
+    if (a->Z_is_one) {
+        if (!FP2_copy(t0, a->Y))
+            goto err;
+    } else {
+        if (!FP2_mul(group, t0, a->Y, a->Z, ctx))
+            goto err;
+    }
+    if (!FP2_dbl(group, r->Z, t0))
+        goto err;
+    r->Z_is_one = 0;
+    /*
+     * Z_r = 2 * Y_a * Z_a
+     */
+
+    /*
+     * t2
+     */
+    if (!FP2_sqr(group, t3, a->Y, ctx))
+        goto err;
+    if (!FP2_mul(group, t2, a->X, t3, ctx))
+        goto err;
+    if (!FP2_dbl(group, t2, t2))
+        goto err;
+    if (!FP2_dbl(group, t2, t2))
+        goto err;
+    /*
+     * t2 = 4 * X_a * Y_a^2
+     */
+
+    /*
+     * X_r
+     */
+    if (!FP2_dbl(group, t0, t2))
+        goto err;
+    if (!FP2_sqr(group, r->X, t1, ctx))
+        goto err;
+    if (!FP2_sub(group, r->X, r->X, t0))
+        goto err;
+    /*
+     * X_r = t1^2 - 2 * t2
+     */
+
+    /*
+     * t3
+     */
+    if (!FP2_sqr(group, t0, t3, ctx))
+        goto err;
+    if (!FP2_dbl(group, t3, t0))
+        goto err;
+    if (!FP2_dbl(group, t3, t3))
+        goto err;
+    if (!FP2_dbl(group, t3, t3))
+        goto err;
+    /*
+     * t3 = 8 * Y_a^4
+     */
+
+    /*
+     * Y_r
+     */
+    if (!FP2_sub(group, t0, t2, r->X))
+        goto err;
+    if (!FP2_mul(group, t0, t1, t0, ctx))
+        goto err;
+    if (!FP2_sub(group, r->Y, t0, t3))
+        goto err;
+    /*
+     * Y_r = t1 * (t2 - X_r) - t3
+     */
+
+    ret = 1;
+
+ err:
+    FP2_free(t0);
+    FP2_free(t1);
+    FP2_free(t2);
+    FP2_free(t3);
+    BN_CTX_end(ctx);
+    BN_CTX_free(new_ctx);
+    return ret;
+}
+
+int G2_ELEM_invert(const BP_GROUP *group, G2_ELEM *point, BN_CTX *ctx)
+{
+    if (G2_ELEM_is_at_infinity(group, point) || FP2_is_zero(point->Y))
+        /*
+         * point is its own inverse
+         */
+        return 1;
+
+    return FP2_neg(group, point->Y, point->Y);
+
+}
+
+int G2_ELEM_is_at_infinity(const BP_GROUP *group, const G2_ELEM *point)
+{
+    return FP2_is_zero(point->Z);
+}
+
+int G2_ELEM_is_on_curve(const BP_GROUP *group, const G2_ELEM *point,
+                        BN_CTX *ctx)
+{
+    BN_CTX *new_ctx = NULL;
+    FP2 *rh, *t = NULL, *z4 = NULL, *z6 = NULL;
+    int ret = -1;
```

Comment:
Yes, you can set a different error, which will be helpful for logging/debugging. But 100% of the calling code for the EC method either checks for `<= 0` or should be doing so but isn't, leading to a bug.